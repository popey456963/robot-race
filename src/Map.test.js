const { createMap, setMapTile, getMapTile, getMapTilesByType, countAllMapTiles, canMoveInDirection } = require("./Map")
const { createHoleTile, createGearTile, createPlainTile } = require("./Tiles")
const { HOLE, PLAIN, GEAR, EAST, OUT_OF_BOUNDS } = require("./Constants")

function setupSmallMap() {
    const map = createMap(2, 2)

    setMapTile(map, { x: 1, y: 1 }, createHoleTile())
    setMapTile(map, { x: 0, y: 1 }, createGearTile())
    setMapTile(map, { x: 1, y: 0 }, createGearTile())

    return map
}

test('can create a map', () => {
    expect(createMap(1, 1)).toMatchSnapshot()
    expect(createMap(2, 3)).toMatchSnapshot()
    expect(createMap(3, 2)).toMatchSnapshot()
})

test('get / set map tile', () => {
    const map = setupSmallMap()

    setMapTile(map, { x: 1, y: 1 }, createHoleTile())
    setMapTile(map, { x: 0, y: 1 }, createGearTile())

    expect(map).toMatchSnapshot()

    expect(getMapTile(map, { x: 1, y: 1 }).type).toBe(HOLE)
    expect(getMapTile(map, { x: 0, y: 1 }).type).toBe(GEAR)
    expect(getMapTile(map, { x: 0, y: 0 }).type).toBe(PLAIN)

    expect(getMapTile(map, { x: 3, y: 0 }).type).toBe(OUT_OF_BOUNDS)
    expect(getMapTile(map, { x: 0, y: 3 }).type).toBe(OUT_OF_BOUNDS)
})

test('get map tiles by type', () => {
    const map = setupSmallMap()

    expect(getMapTilesByType(map, GEAR)).toMatchSnapshot()
    expect(getMapTilesByType(map, HOLE)).toMatchSnapshot()
    expect(getMapTilesByType(map, PLAIN)).toMatchSnapshot()
})

test('count tiles in map', () => {
    const map = setupSmallMap()

    expect(countAllMapTiles(map)).toMatchSnapshot()
})

test('can move in direction', () => {
    const map = createMap(2, 2)

    const plain = createPlainTile({ walls: [EAST] })
    setMapTile(map, { x: 0, y: 0 }, plain)

    expect(canMoveInDirection(map, { x: 0, y: 0 }, EAST)).toBe(false)
    expect(canMoveInDirection(map, { x: 0, y: 1 }, EAST)).toBe(true)
})