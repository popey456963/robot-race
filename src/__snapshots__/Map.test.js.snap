// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a map 1`] = `
Array [
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
]
`;

exports[`can create a map 2`] = `
Array [
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 1,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 1,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 2,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 2,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
]
`;

exports[`can create a map 3`] = `
Array [
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 2,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 1,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 1,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 2,
        "y": 1,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
  ],
]
`;

exports[`count tiles in map 1`] = `
Object {
  "GEAR": 2,
  "HOLE": 1,
  "PLAIN": 1,
}
`;

exports[`get / set map tile 1`] = `
Array [
  Array [
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "pushers": Object {},
      "type": "PLAIN",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "meta": Object {
        "rotationDirection": undefined,
      },
      "position": Object {
        "x": 1,
        "y": 0,
      },
      "pushers": Object {},
      "type": "GEAR",
      "walls": Object {},
    },
  ],
  Array [
    Object {
      "lasers": Object {},
      "meta": Object {
        "rotationDirection": undefined,
      },
      "position": Object {
        "x": 0,
        "y": 1,
      },
      "pushers": Object {},
      "type": "GEAR",
      "walls": Object {},
    },
    Object {
      "lasers": Object {},
      "position": Object {
        "x": 1,
        "y": 1,
      },
      "pushers": Object {},
      "type": "HOLE",
      "walls": Object {},
    },
  ],
]
`;

exports[`get map tiles by type 1`] = `
Array [
  Object {
    "lasers": Object {},
    "meta": Object {
      "rotationDirection": undefined,
    },
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "pushers": Object {},
    "type": "GEAR",
    "walls": Object {},
  },
  Object {
    "lasers": Object {},
    "meta": Object {
      "rotationDirection": undefined,
    },
    "position": Object {
      "x": 0,
      "y": 1,
    },
    "pushers": Object {},
    "type": "GEAR",
    "walls": Object {},
  },
]
`;

exports[`get map tiles by type 2`] = `
Array [
  Object {
    "lasers": Object {},
    "position": Object {
      "x": 1,
      "y": 1,
    },
    "pushers": Object {},
    "type": "HOLE",
    "walls": Object {},
  },
]
`;

exports[`get map tiles by type 3`] = `
Array [
  Object {
    "lasers": Object {},
    "position": Object {
      "x": 0,
      "y": 0,
    },
    "pushers": Object {},
    "type": "PLAIN",
    "walls": Object {},
  },
]
`;
